servers/zms/src/main/java/com/yahoo/athenz/zms/store/jdbc/JDBCConnection.java

```diff
@@ -3355,6 +3363,89 @@ public DomainRoleMembers listDomainRoleMembers(String domainName) {
         return domainRoleMembers;
     }
 
+
+    @Override
+    public boolean confirmRoleMember(String domainName, String roleName, RoleMember roleMember,
+                                    String admin, String auditRef) {
+
+        final String caller = "confirmRoleMember";
+
+        String principal = roleMember.getMemberName();
+        int domainId = getDomainId(domainName);
+        if (domainId == 0) {
+            throw notFoundError(caller, ZMSConsts.OBJECT_DOMAIN, domainName);
+        }
+        int roleId = getRoleId(domainId, roleName);
+        if (roleId == 0) {
+            throw notFoundError(caller, ZMSConsts.OBJECT_ROLE, ZMSUtils.roleResourceName(domainName, roleName));
+        }
+        if (!validatePrincipalDomain(principal)) {
+            throw notFoundError(caller, ZMSConsts.OBJECT_DOMAIN, principal);
+        }
+        int principalId = getPrincipalId(principal);
+        if (principalId == 0) {
+            principalId = insertPrincipal(principal);
+            if (principalId == 0) {
+                throw internalServerError(caller, "Unable to insert principal: " + principal);
+            }
+        }
+        //need to check if entry already exists
+        int affectedRows;
+        boolean roleMemberExists = false;
+        boolean result;
+        try (PreparedStatement ps = con.prepareStatement(SQL_GET_ROLE_MEMBER_INACTIVE_EXISTS)) {
+            ps.setInt(1, roleId);
+            ps.setInt(2, principalId);
+            try (ResultSet rs = executeQuery(ps, caller)) {
+                if (rs.next()) {
+                    roleMemberExists = true;
+                }
+            }
+        } catch (SQLException ex) {
+            throw sqlError(ex, caller);
+        }
+        java.sql.Timestamp expiration = null;
```

Comment:
at this time we should check if roleMemberExists or not and not on line 3411. if it doesn't exist, then throw an exception but not an internal error - it's a 404 so a not found error